# CREDIT CARD FRAUD DETECTION

In [2]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

from sklearn.metrics import accuracy_score

In [79]:
data= pd.read_csv("/content/creditcard.csv",error_bad_lines=False)

<ipython-input-79-094079499662>:1: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.


  data= pd.read_csv("/content/creditcard.csv",error_bad_lines=False)
Skipping line 3974: expected 31 fields, saw 40



In [80]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 286793 entries, 0 to 286792
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    286793 non-null  float64
 1   V1      286793 non-null  float64
 2   V2      286793 non-null  float64
 3   V3      286793 non-null  float64
 4   V4      286793 non-null  float64
 5   V5      286793 non-null  float64
 6   V6      286793 non-null  float64
 7   V7      286793 non-null  float64
 8   V8      286793 non-null  float64
 9   V9      286793 non-null  float64
 10  V10     286793 non-null  float64
 11  V11     286793 non-null  float64
 12  V12     286793 non-null  float64
 13  V13     286793 non-null  float64
 14  V14     286793 non-null  float64
 15  V15     286793 non-null  float64
 16  V16     286793 non-null  float64
 17  V17     286793 non-null  float64
 18  V18     286793 non-null  float64
 19  V19     286793 non-null  float64
 20  V20     286793 non-null  float64
 21  V21     28

In [81]:
#null values are present or not
data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [82]:
data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [83]:
#Dropping null values
data.dropna(how='any', inplace=True)

In [84]:
data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [85]:
data['Class'].value_counts()

0    286301
1       492
Name: Class, dtype: int64

In [86]:
#Handling Class Imbalance
legit_transactions=data[data.Class==0]
fraud_transactions=data[data.Class==1]


In [87]:
print(legit_transactions.shape)
print(fraud_transactions.shape)

(286301, 31)
(492, 31)


In [88]:
legit_transactions.Amount.describe()

count    286301.000000
mean         88.104564
std         249.703884
min           0.000000
25%           5.500000
50%          21.990000
75%          76.970000
max       25691.160000
Name: Amount, dtype: float64

In [89]:
fraud_transactions.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [90]:
data.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94197.781653  0.005543 -0.003799  0.017815 -0.008674  0.005547   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002787  0.010646 -0.001562  0.005120  ... -0.000207 -0.000999   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000171 -0.000444  0.000487  0.000501 -0.000606  0.000137 -0.000080   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.104564  
1      122.211321  

[2 rows x 30 columns]

In [96]:
# UNDERSAMPLING
legit_sample=legit_transactions.sample(n=492)
dataset=pd.concat([legit_sample,fraud_transactions])
dataset.head()

Time        V1        V2        V3        V4        V5        V6  \
152684   93922.0  1.916981 -0.284806 -0.095547  0.739449 -0.609655 -0.437915   
117895   74068.0  1.126291 -1.407402  1.903831 -0.279499 -1.826804  1.675483   
219928  141103.0  1.067690 -3.587023 -1.525961 -0.868122 -1.954332  0.131116   
270464  163231.0  0.093168  0.966579 -0.337094 -0.623622  0.868786 -0.590852   
47796    42506.0 -0.575971  0.162415  0.688628 -0.400275  2.567916  3.854092   

              V7        V8        V9  ...       V21       V22       V23  \
152684 -0.568521 -0.219332  2.517841  ...  0.139924  0.942253  0.057241   
117895 -2.106987  0.765377  0.831846  ...  0.375214  1.532179 -0.126271   
219928 -0.557276 -0.176052 -0.880851  ...  0.457939  0.724345 -0.631089   
270464  0.864292  0.043686 -0.271917  ... -0.283960 -0.722863  0.106844   
47796   0.003093  0.800844 -0.296996  ... -0.305292 -0.838067 -0.148919   

             V24       V25       V26       V27       V28  Amount  Class  
152684  0.005775 -0.041684 -0.198077  0.023272 -0.036037   37.89      0  
117895 -0.181696  0.346557  0.138794  0.131973  0.014849   18.50      0  
219928 -0.310172  0.040193  0.190454 -0.090471  0.032767  628.00      0  
270464  0.652280 -0.477954  0.087590  0.216166  0.074949    5.16      0  
47796   1.023859  0.277927  0.347601 -0.129558 -0.148248   34.99      0  

[5 rows x 31 columns]

In [97]:
dataset['Class'].value_counts()

0    492
1    492
Name: Class, dtype: int64

In [98]:
x

Time        V1        V2        V3        V4        V5        V6  \
3650  3121  1.292089  0.236557  0.071134  0.320282  0.282772  0.086503   
985    744  0.846466 -0.648165  0.909723  1.278921 -0.848055  0.723489   
1993  1531 -2.844709  2.734210 -0.499037 -1.034623 -0.776265 -0.668750   
41      32 -2.008872  2.198527  0.144242  1.159432 -0.815174  0.182288   
1753  1355 -4.554545  1.419867  0.212262 -2.341553 -2.623821  2.566345   
...    ...       ...       ...       ...       ...       ...       ...   
3006  2555 -0.788138  0.619180  1.568469 -0.913697  0.378057  0.157612   
848    646 -0.691375  0.348838  1.792883 -0.178102  1.057957  1.780065   
3803  3316 -0.358519  0.650343  1.082354 -2.026716  0.097420 -1.522650   
541    406 -2.312227  1.951992 -1.609851  3.997906 -0.522188 -1.426545   
623    472 -3.043541 -3.157307  1.088463  2.288644  1.359805 -1.064823   

            V7         V8        V9  ...       V20       V21       V22  \
3650  0.016469  -0.047239 -0.011877  ... -0.040991 -0.288744 -0.752383   
985  -0.706530   0.453813  0.762523  ... -0.002306  0.238578  0.464747   
1993 -0.356853   1.313071  0.591929  ...  0.536069 -0.289244 -0.643686   
41   -0.617108   1.530817 -0.586832  ...  0.028740  0.094917  0.294983   
1753 -4.533102 -11.285105  0.743721  ...  0.635598 -1.184517  0.905380   
...        ...        ...       ...  ...       ...       ...       ...   
3006  0.391596   0.133121 -0.457557  ...  0.065520 -0.076895 -0.388709   
848   0.520891   0.208006  0.111160  ... -0.145018 -0.069795  0.333184   
3803  1.115724  -0.394219  0.963631  ... -0.024650  0.050297  0.483723   
541  -2.537387   1.391657 -2.770089  ...  0.126911  0.517232 -0.035049   
623   0.325574  -0.067794 -0.270953  ...  2.102339  0.661696  0.435477   

           V23       V24       V25       V26       V27       V28  Amount  
3650  0.029499 -0.942518  0.310273  0.185592 -0.008402  0.006752    0.89  
985  -0.204267 -0.327269  0.330811 -0.217592  0.041951  0.037855  138.00  
1993  0.027395 -0.432674  0.296821  0.376164  0.556133  0.352268    1.00  
41    0.011081  0.015249  0.034211 -0.236141  0.128291  0.117986    2.35  
1753  0.594792 -0.029651 -0.615249  0.153999  0.186242  0.653324  200.00  
...        ...       ...       ...       ...       ...       ...     ...  
3006 -0.336370 -0.809746  0.187735  0.749377 -0.162004  0.060985   22.89  
848   0.000205 -1.008831 -0.425738  0.262964 -0.492098 -0.423527    1.18  
3803 -0.286300  0.362213  0.044769 -0.836287  0.233261 -0.015568    1.00  
541  -0.465211  0.320198  0.044519  0.177840  0.261145 -0.143276    0.00  
623   1.375966 -0.293803  0.279798 -0.145362 -0.252773  0.035764  529.00  

[984 rows x 30 columns]

In [99]:
dataset.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      91605.853659 -0.043956 -0.061500  0.107749 -0.068133 -0.022871   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.097648 -0.034924  0.091006  0.044321  ...  0.015492  0.016393   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.026841 -0.009128  0.073373  0.004416  0.027626  0.029881  0.008402   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       93.165915  
1      122.211321  

[2 rows x 30 columns]

In [100]:
x=dataset.drop(columns='Class',axis=1)
y=dataset['Class']

In [101]:
y

152684    0
117895    0
219928    0
270464    0
47796     0
         ..
281849    1
282129    1
282135    1
283130    1
283660    1
Name: Class, Length: 984, dtype: int64

#Test and Train Dataset

In [102]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,stratify=y,random_state=2)
print(x_train.shape,x_test.shape,x.shape)

(787, 30) (197, 30) (984, 30)


In [103]:
# MODEL TRAINING
model= LogisticRegression()
print(model)

LogisticRegression()


In [105]:
model.fit(x_train,y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [106]:
x_train_pred=model.predict(x_train)
training_data_accuracy=accuracy_score(x_train_pred,y_train)

In [107]:
print("Accuracy of the model:",training_data_accuracy*100)

Accuracy of the model: 95.80686149936467


In [108]:
# LET US CALUCULATE THE ACCURACY SCORE ON TESTING DATA SAMPLE
x_test_pred=model.predict(x_test)
testing_accuracy=accuracy_score(x_test_pred,y_test)
print("Accuracy of the model with testing Data:",testing_accuracy*100)

Accuracy of the model with testing Data: 94.9238578680203


In [110]:
# CALCULATING PRECISION, RECALL AND F1 SCORE
from sklearn.metrics import precision_score, recall_score,f1_score


In [114]:
#PRECISION

precision = precision_score(x_test_pred,y_test)
print("Precision of the model:",precision)

Precision of the model: 0.9081632653061225


In [112]:
#RECALL

recall = recall_score(x_test_pred,y_test)
print("Recall of the model:",recall )

Recall of the model: 0.9888888888888889


In [113]:
#F1

f1 = f1_score(x_test_pred,y_test)
print("F1 Score of the model:",f1)

F1 Score of the model: 0.946808510638298


Precision of the model: 0.9081632653061225

Recall of the model: 0.9888888888888889

F1 Score of the model: 0.946808510638298

HENCE WE HAVE CREATED A MODEL USING LOGISTIC REGRESSION, TRAINED THEM USING TRAIN DATA AND TESTED THEM USING TESTING DATA WHICH HAS YIELDED US A ACCURACY OF 94%